ta/pkcs11/src/persistent_token.c

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <pkcs11_ta.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_internal_api_extensions.h>
+#include <util.h>
+
+#include "pkcs11_token.h"
+#include "pkcs11_helpers.h"
+
+/* 'X' will be replaced by the token decimal id (up to 9!) */
+#define TOKEN_DB_FILE_BASE		"token.db.X"
+
+void close_persistent_db(struct ck_token *token)
+{
+	int n = 0;
+
+	for (n = 0; n < PKCS11_MAX_USERS; n++) {
+		TEE_CloseObject(token->pin_hdl[n]);
+		token->pin_hdl[n] = TEE_HANDLE_NULL;
+	}
+
+	TEE_CloseObject(token->db_hdl);
+	token->db_hdl = TEE_HANDLE_NULL;
+}
+
+static void init_pin_keys(struct ck_token *token, unsigned int uid)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int token_id = get_token_id(token);
+	TEE_ObjectHandle *key_hdl = &token->pin_hdl[uid];
+	char file[32] = { 0 };
+
+	assert(token_id < 10 && uid < 10);
+
+	if (snprintf(file, 32, "token.db.%1d-pin%1d", token_id, uid) >= 32)
+		TEE_Panic(0);
+
+	res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE,
+				       file, sizeof(file), 0, key_hdl);
+	if (res == TEE_SUCCESS) {
+		DMSG("PIN key found");
+		return;
+	}
+
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		TEE_Attribute attr = { };
+		TEE_ObjectHandle hdl = TEE_HANDLE_NULL;
+		uint8_t pin_key[16] = { 0 };
+
+		TEE_MemFill(&attr, 0, sizeof(attr));
+
+		TEE_GenerateRandom(pin_key, sizeof(pin_key));
+		TEE_InitRefAttribute(&attr, TEE_ATTR_SECRET_VALUE,
+				     pin_key, sizeof(pin_key));
+
+		res = TEE_AllocateTransientObject(TEE_TYPE_AES, 128, key_hdl);
+		if (res)
+			TEE_Panic(0);
+
+		res = TEE_PopulateTransientObject(*key_hdl, &attr, 1);
+		if (res)
+			TEE_Panic(0);
+
+		res = TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE,
+						 file, sizeof(file), 0,
+						 *key_hdl,
+						 pin_key, sizeof(pin_key),
+						 &hdl);
+		if (res)
+			TEE_Panic(0);
+
+		TEE_CloseObject(hdl);
+		return;
+	}
+
+	TEE_Panic(0);
+}
+
+/*
+ * Return the token instance, either initialized from reset or initialized
+ * from the token persistent state if found.
+ */
+struct ck_token *init_persistent_db(unsigned int token_id)
```

Comment:
This function will use a bit of the TEE Core heap by having multiple storage object open all the time.
How about only opening the storage object when needed?
If something is performance critical one could consider keeping some object open while it's needed. But I think that's better to add at then end when we can see that it really makes a difference.

Reply:
Fair. Then I change `close_persistent_db()` into an empty function for now, so that it' sill called to balance the call to  `init_persistent_db()`.